src/fairseq2/utils/read_and_pickle_wrap_iterator.py

```diff
@@ -0,0 +1,48 @@
+from typing import Callable, Iterator, Tuple, TypeVar
+
+from typing_extensions import Self
+
+from fairseq2.data import DataPipelineBuilder, read_iterator
+
+T = TypeVar("T")
+
+
+class IteratorPickleWrapper(Iterator[T]):
+    def __init__(self, iterator_factory: Callable[[], Iterator[T]]) -> None:
+        self.iterator_factory = iterator_factory
```

Comment:
I suggest prefixing the member variables with an underscore to indicate that they are private state:

```python
self._iterator_factory = ...
self._iterator = ...
self._counter = ...
```